{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from numpy import ones\n",
    "from numpy_sugar import ddot\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "from pandas_plink import read_plink1_bin\n",
    "from numpy.linalg import cholesky\n",
    "from numpy_sugar.linalg import economic_svd\n",
    "import xarray as xr\n",
    "from struct_lmm2 import StructLMM2\n",
    "from limix.qc import quantile_gaussianize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# in the actual script this will be provided as an argument\n",
    "chrom = 22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_files_dir = \"/hps/nobackup/stegle/users/acuomo/all_scripts/struct_LMM2/sc_endodiff/new/input_files/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype_individual_id</th>\n",
       "      <th>phenotype_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HPSI0114i-joxm_1</td>\n",
       "      <td>21843_1#10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HPSI0314i-fafq_1</td>\n",
       "      <td>21843_1#100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HPSI0314i-fafq_1</td>\n",
       "      <td>21843_1#101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HPSI1013i-wuye_2</td>\n",
       "      <td>21843_1#102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HPSI0114i-joxm_1</td>\n",
       "      <td>21843_1#103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  genotype_individual_id phenotype_sample_id\n",
       "0       HPSI0114i-joxm_1          21843_1#10\n",
       "1       HPSI0314i-fafq_1         21843_1#100\n",
       "2       HPSI0314i-fafq_1         21843_1#101\n",
       "3       HPSI1013i-wuye_2         21843_1#102\n",
       "4       HPSI0114i-joxm_1         21843_1#103"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## this file will map cells to donors, it will also only including donors we have single cell data (a subset of all of HipSci donors)\n",
    "sample_mapping_file = input_files_dir+\"sample_mapping_file.csv\"\n",
    "sample_mapping = pd.read_csv(sample_mapping_file, dtype={\"genotype_individual_id\": str, \"phenotype_sample_id\": str})\n",
    "sample_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique donors: 126\n"
     ]
    }
   ],
   "source": [
    "## extract unique individuals\n",
    "donors = sample_mapping[\"genotype_individual_id\"].unique()\n",
    "donors.sort()\n",
    "print(\"Number of unique donors: {}\".format(len(donors)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [05:53<00:00, 117.77s/it]\n"
     ]
    }
   ],
   "source": [
    "## read in genotype file (plink format)\n",
    "plink_file = \"/hps/nobackup/hipsci/scratch/genotypes/imputed/2017-03-27/Full_Filtered_SNPs_Plink/hipsci.wec.gtarray.HumanCoreExome.imputed_phased.20170327.genotypes.norm.renamed.bed\"\n",
    "G = read_plink1_bin(plink_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of donors after kinship intersection: 125\n"
     ]
    }
   ],
   "source": [
    "## read in GRM kinship matrix\n",
    "kinship_file = \"/hps/nobackup/hipsci/scratch/genotypes/imputed/2017-03-27/Full_Filtered_SNPs_Plink-F/hipsci.wec.gtarray.HumanCoreExome.imputed_phased.20170327.genotypes.norm.renamed.kinship\"\n",
    "K = pd.read_csv(kinship_file, sep=\"\\t\", index_col=0)\n",
    "assert all(K.columns == K.index)\n",
    "K = xr.DataArray(K.values, dims=[\"sample_0\", \"sample_1\"], coords={\"sample_0\": K.columns, \"sample_1\": K.index})\n",
    "K = K.sortby(\"sample_0\").sortby(\"sample_1\")\n",
    "donors = sorted(set(list(K.sample_0.values)).intersection(donors))\n",
    "print(\"Number of donors after kinship intersection: {}\".format(len(donors)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## subset to relevant donors (from sample mapping file)\n",
    "K = K.sel(sample_0=donors, sample_1=donors)\n",
    "assert all(K.sample_0 == donors)\n",
    "assert all(K.sample_1 == donors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## and decompose such that K = L @ L.T\n",
    "L_kinship = cholesky(K.values)\n",
    "L_kinship = xr.DataArray(L_kinship, dims=[\"sample\", \"col\"], coords={\"sample\": K.sample_0.values})\n",
    "assert all(L_kinship.sample.values == K.sample_0.values)\n",
    "del K"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample mapping number of rows BEFORE intersection: 34256\n",
      "Sample mapping number of rows AFTER intersection: 33964\n"
     ]
    }
   ],
   "source": [
    "# number of samples (cells)\n",
    "print(\"Sample mapping number of rows BEFORE intersection: {}\".format(sample_mapping.shape[0]))\n",
    "sample_mapping = sample_mapping[sample_mapping[\"genotype_individual_id\"].isin(donors)]\n",
    "print(\"Sample mapping number of rows AFTER intersection: {}\".format(sample_mapping.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# expand from donors to cells\n",
    "L_expanded = L_kinship.sel(sample=sample_mapping[\"genotype_individual_id\"].values)\n",
    "assert all(L_expanded.sample.values == sample_mapping[\"genotype_individual_id\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# environments\n",
    "# cells by PCs (20)\n",
    "E_file = input_files_dir+\"20PCs.csv\" \n",
    "E = pd.read_csv(E_file, index_col = 0)\n",
    "E = xr.DataArray(E.values, dims=[\"cell\", \"pc\"], coords={\"cell\": E.index.values, \"pc\": E.columns.values})\n",
    "E = E.sel(cell=sample_mapping[\"phenotype_sample_id\"].values)\n",
    "assert all(E.cell.values == sample_mapping[\"phenotype_sample_id\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subselect to only SNPs on right chromosome\n",
    "G_sel = G.where(G.chrom == str(chrom), drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2 {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.DataArray &#x27;genotype&#x27; (sample: 1610, variant: 131988)&gt;\n",
       "dask.array&lt;where, shape=(1610, 131988), dtype=float32, chunksize=(1024, 1024), chunktype=numpy.ndarray&gt;\n",
       "Coordinates:\n",
       "  * sample   (sample) object &#x27;HPSI0216i-tony_5&#x27; ... &#x27;HPSI0416pf-cihd&#x27;\n",
       "  * variant  (variant) object &#x27;22_22_16051249_T_C&#x27; ... &#x27;22_22_51229805_T_C&#x27;\n",
       "    fid      (sample) object &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; ... &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27;\n",
       "    iid      (sample) object &#x27;HPSI0216i-tony_5&#x27; ... &#x27;HPSI0416pf-cihd&#x27;\n",
       "    father   (sample) object &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; ... &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27;\n",
       "    mother   (sample) object &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; ... &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27;\n",
       "    gender   (sample) object &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; ... &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27;\n",
       "    trait    (sample) object &#x27;-9&#x27; &#x27;-9&#x27; &#x27;-9&#x27; &#x27;-9&#x27; &#x27;-9&#x27; ... &#x27;-9&#x27; &#x27;-9&#x27; &#x27;-9&#x27; &#x27;-9&#x27;\n",
       "    chrom    (variant) object &#x27;22&#x27; &#x27;22&#x27; &#x27;22&#x27; &#x27;22&#x27; &#x27;22&#x27; ... &#x27;22&#x27; &#x27;22&#x27; &#x27;22&#x27; &#x27;22&#x27;\n",
       "    snp      (variant) object &#x27;22_16051249_T_C&#x27; ... &#x27;22_51229805_T_C&#x27;\n",
       "    cm       (variant) float64 0.0 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0 0.0\n",
       "    pos      (variant) int32 16051249 16051453 16053862 ... 51228910 51229805\n",
       "    a0       (variant) object &#x27;T&#x27; &#x27;A&#x27; &#x27;C&#x27; &#x27;C&#x27; &#x27;G&#x27; &#x27;C&#x27; ... &#x27;G&#x27; &#x27;T&#x27; &#x27;A&#x27; &#x27;G&#x27; &#x27;T&#x27;\n",
       "    a1       (variant) object &#x27;C&#x27; &#x27;C&#x27; &#x27;T&#x27; &#x27;T&#x27; &#x27;T&#x27; &#x27;T&#x27; ... &#x27;A&#x27; &#x27;C&#x27; &#x27;G&#x27; &#x27;A&#x27; &#x27;C&#x27;</pre><div class='xr-wrap' hidden><div class='xr-header'><div class='xr-obj-type'>xarray.DataArray</div><div class='xr-array-name'>'genotype'</div><ul class='xr-dim-list'><li><span class='xr-has-index'>sample</span>: 1610</li><li><span class='xr-has-index'>variant</span>: 131988</li></ul></div><ul class='xr-sections'><li class='xr-section-item'><div class='xr-array-wrap'><input id='section-36d16152-bcb6-4bc6-9343-9d1e773aff66' class='xr-array-in' type='checkbox' checked><label for='section-36d16152-bcb6-4bc6-9343-9d1e773aff66' title='Show/hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-array-preview xr-preview'><span>dask.array&lt;chunksize=(1024, 45), meta=np.ndarray&gt;</span></div><div class='xr-array-data'><table>\n",
       "<tr>\n",
       "<td>\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr><td> </td><th> Array </th><th> Chunk </th></tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr><th> Bytes </th><td> 850.00 MB </td> <td> 4.19 MB </td></tr>\n",
       "    <tr><th> Shape </th><td> (1610, 131988) </td> <td> (1024, 1024) </td></tr>\n",
       "    <tr><th> Count </th><td> 102981 Tasks </td><td> 260 Chunks </td></tr>\n",
       "    <tr><th> Type </th><td> float32 </td><td> numpy.ndarray </td></tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</td>\n",
       "<td>\n",
       "<svg width=\"170\" height=\"76\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"16\" x2=\"120\" y2=\"16\" />\n",
       "  <line x1=\"0\" y1=\"26\" x2=\"120\" y2=\"26\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"26\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"4\" y1=\"0\" x2=\"4\" y2=\"26\" />\n",
       "  <line x1=\"11\" y1=\"0\" x2=\"11\" y2=\"26\" />\n",
       "  <line x1=\"17\" y1=\"0\" x2=\"17\" y2=\"26\" />\n",
       "  <line x1=\"24\" y1=\"0\" x2=\"24\" y2=\"26\" />\n",
       "  <line x1=\"30\" y1=\"0\" x2=\"30\" y2=\"26\" />\n",
       "  <line x1=\"37\" y1=\"0\" x2=\"37\" y2=\"26\" />\n",
       "  <line x1=\"42\" y1=\"0\" x2=\"42\" y2=\"26\" />\n",
       "  <line x1=\"49\" y1=\"0\" x2=\"49\" y2=\"26\" />\n",
       "  <line x1=\"55\" y1=\"0\" x2=\"55\" y2=\"26\" />\n",
       "  <line x1=\"62\" y1=\"0\" x2=\"62\" y2=\"26\" />\n",
       "  <line x1=\"68\" y1=\"0\" x2=\"68\" y2=\"26\" />\n",
       "  <line x1=\"75\" y1=\"0\" x2=\"75\" y2=\"26\" />\n",
       "  <line x1=\"81\" y1=\"0\" x2=\"81\" y2=\"26\" />\n",
       "  <line x1=\"87\" y1=\"0\" x2=\"87\" y2=\"26\" />\n",
       "  <line x1=\"94\" y1=\"0\" x2=\"94\" y2=\"26\" />\n",
       "  <line x1=\"100\" y1=\"0\" x2=\"100\" y2=\"26\" />\n",
       "  <line x1=\"107\" y1=\"0\" x2=\"107\" y2=\"26\" />\n",
       "  <line x1=\"113\" y1=\"0\" x2=\"113\" y2=\"26\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"26\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 120.0,0.0 120.0,26.389108936312358 0.0,26.389108936312358\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"46.389109\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >131988</text>\n",
       "  <text x=\"140.000000\" y=\"13.194554\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,140.000000,13.194554)\">1610</text>\n",
       "</svg>\n",
       "</td>\n",
       "</tr>\n",
       "</table></div></div></li><li class='xr-section-item'><input id='section-c3e1b481-514b-4b22-9773-de67c2ee9e07' class='xr-section-summary-in' type='checkbox'  checked><label for='section-c3e1b481-514b-4b22-9773-de67c2ee9e07' class='xr-section-summary' >Coordinates: <span>(14)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample</span></div><div class='xr-var-dims'>(sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;HPSI0216i-tony_5&#x27; ... &#x27;HPSI0416...</div><input id='attrs-4d45c99a-6af7-4838-8bc9-f156e5cc6165' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4d45c99a-6af7-4838-8bc9-f156e5cc6165' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-681213d8-2f39-4494-9806-c6a462fc0ba0' class='xr-var-data-in' type='checkbox'><label for='data-681213d8-2f39-4494-9806-c6a462fc0ba0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;HPSI0216i-tony_5&#x27;, &#x27;HPSI0216pf-tony&#x27;, &#x27;HPSI0316i-xaqm_4&#x27;, ...,\n",
       "       &#x27;HPSI0516pf-eosr&#x27;, &#x27;HPSI0616pf-xuvy&#x27;, &#x27;HPSI0416pf-cihd&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>variant</span></div><div class='xr-var-dims'>(variant)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;22_22_16051249_T_C&#x27; ... &#x27;22_22_...</div><input id='attrs-1ec42663-85e4-457a-8a71-9e62da015320' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1ec42663-85e4-457a-8a71-9e62da015320' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-dbcb7365-7dbe-4a74-be49-9bc0ae788ebc' class='xr-var-data-in' type='checkbox'><label for='data-dbcb7365-7dbe-4a74-be49-9bc0ae788ebc' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;22_22_16051249_T_C&#x27;, &#x27;22_22_16051453_A_C&#x27;, &#x27;22_22_16053862_C_T&#x27;, ...,\n",
       "       &#x27;22_22_51228259_A_G&#x27;, &#x27;22_22_51228910_G_A&#x27;, &#x27;22_22_51229805_T_C&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>fid</span></div><div class='xr-var-dims'>(sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; ... &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27;</div><input id='attrs-486ce4a3-fde0-4be7-88b5-dc87c684658f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-486ce4a3-fde0-4be7-88b5-dc87c684658f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7674c8be-3e38-4435-ae68-52e2a18734db' class='xr-var-data-in' type='checkbox'><label for='data-7674c8be-3e38-4435-ae68-52e2a18734db' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;0&#x27;, &#x27;0&#x27;, &#x27;0&#x27;, ..., &#x27;0&#x27;, &#x27;0&#x27;, &#x27;0&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>iid</span></div><div class='xr-var-dims'>(sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;HPSI0216i-tony_5&#x27; ... &#x27;HPSI0416...</div><input id='attrs-97e5d33f-79c8-47c2-95e2-9404f6eeadf7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-97e5d33f-79c8-47c2-95e2-9404f6eeadf7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d8233c1e-fe0b-448a-891a-001d25898149' class='xr-var-data-in' type='checkbox'><label for='data-d8233c1e-fe0b-448a-891a-001d25898149' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;HPSI0216i-tony_5&#x27;, &#x27;HPSI0216pf-tony&#x27;, &#x27;HPSI0316i-xaqm_4&#x27;, ...,\n",
       "       &#x27;HPSI0516pf-eosr&#x27;, &#x27;HPSI0616pf-xuvy&#x27;, &#x27;HPSI0416pf-cihd&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>father</span></div><div class='xr-var-dims'>(sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; ... &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27;</div><input id='attrs-5732d3ed-e86c-4808-b5ea-d55741b37933' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5732d3ed-e86c-4808-b5ea-d55741b37933' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c8b7df43-e1aa-4147-958b-bd466a87c95a' class='xr-var-data-in' type='checkbox'><label for='data-c8b7df43-e1aa-4147-958b-bd466a87c95a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;0&#x27;, &#x27;0&#x27;, &#x27;0&#x27;, ..., &#x27;0&#x27;, &#x27;0&#x27;, &#x27;0&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>mother</span></div><div class='xr-var-dims'>(sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; ... &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27;</div><input id='attrs-0317d20d-3ffb-4a9b-8d76-877adf671689' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0317d20d-3ffb-4a9b-8d76-877adf671689' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c5d94d50-fd7d-4bee-8f90-de8b1f30c0f6' class='xr-var-data-in' type='checkbox'><label for='data-c5d94d50-fd7d-4bee-8f90-de8b1f30c0f6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;0&#x27;, &#x27;0&#x27;, &#x27;0&#x27;, ..., &#x27;0&#x27;, &#x27;0&#x27;, &#x27;0&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>gender</span></div><div class='xr-var-dims'>(sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; ... &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27; &#x27;0&#x27;</div><input id='attrs-2e91ec22-4811-4ffb-acdd-39b95d831c7f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2e91ec22-4811-4ffb-acdd-39b95d831c7f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-31676be2-1e89-4f4a-987b-25f1bf8d02db' class='xr-var-data-in' type='checkbox'><label for='data-31676be2-1e89-4f4a-987b-25f1bf8d02db' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;0&#x27;, &#x27;0&#x27;, &#x27;0&#x27;, ..., &#x27;0&#x27;, &#x27;0&#x27;, &#x27;0&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>trait</span></div><div class='xr-var-dims'>(sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;-9&#x27; &#x27;-9&#x27; &#x27;-9&#x27; ... &#x27;-9&#x27; &#x27;-9&#x27; &#x27;-9&#x27;</div><input id='attrs-a6833be5-90cb-439c-adb4-e2078792b1a8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a6833be5-90cb-439c-adb4-e2078792b1a8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-bea78743-6947-4620-8770-c53d5ecf1771' class='xr-var-data-in' type='checkbox'><label for='data-bea78743-6947-4620-8770-c53d5ecf1771' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;-9&#x27;, &#x27;-9&#x27;, &#x27;-9&#x27;, ..., &#x27;-9&#x27;, &#x27;-9&#x27;, &#x27;-9&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>chrom</span></div><div class='xr-var-dims'>(variant)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;22&#x27; &#x27;22&#x27; &#x27;22&#x27; ... &#x27;22&#x27; &#x27;22&#x27; &#x27;22&#x27;</div><input id='attrs-033149fd-692a-4e7b-b7ee-54ab571589e1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-033149fd-692a-4e7b-b7ee-54ab571589e1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-80774e36-5add-4e8e-b66e-9cebd6f3ab26' class='xr-var-data-in' type='checkbox'><label for='data-80774e36-5add-4e8e-b66e-9cebd6f3ab26' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;22&#x27;, &#x27;22&#x27;, &#x27;22&#x27;, ..., &#x27;22&#x27;, &#x27;22&#x27;, &#x27;22&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>snp</span></div><div class='xr-var-dims'>(variant)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;22_16051249_T_C&#x27; ... &#x27;22_512298...</div><input id='attrs-e7a0e231-6243-44af-b7ef-68c0af57bd8b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e7a0e231-6243-44af-b7ef-68c0af57bd8b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9017dfc0-64c3-4df9-a5c9-c9ea828f7c5f' class='xr-var-data-in' type='checkbox'><label for='data-9017dfc0-64c3-4df9-a5c9-c9ea828f7c5f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;22_16051249_T_C&#x27;, &#x27;22_16051453_A_C&#x27;, &#x27;22_16053862_C_T&#x27;, ...,\n",
       "       &#x27;22_51228259_A_G&#x27;, &#x27;22_51228910_G_A&#x27;, &#x27;22_51229805_T_C&#x27;],\n",
       "      dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>cm</span></div><div class='xr-var-dims'>(variant)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0</div><input id='attrs-3c5590ae-63f7-4525-9e37-39d0d32ba408' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3c5590ae-63f7-4525-9e37-39d0d32ba408' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5b650c42-6602-444d-8c5c-ad1d1e891edc' class='xr-var-data-in' type='checkbox'><label for='data-5b650c42-6602-444d-8c5c-ad1d1e891edc' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([0., 0., 0., ..., 0., 0., 0.])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>pos</span></div><div class='xr-var-dims'>(variant)</div><div class='xr-var-dtype'>int32</div><div class='xr-var-preview xr-preview'>16051249 16051453 ... 51229805</div><input id='attrs-29b46607-100e-453e-bb5f-a89d4910ff39' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-29b46607-100e-453e-bb5f-a89d4910ff39' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f29a4bbc-160d-4fed-9dd3-718a2bb07a21' class='xr-var-data-in' type='checkbox'><label for='data-f29a4bbc-160d-4fed-9dd3-718a2bb07a21' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([16051249, 16051453, 16053862, ..., 51228259, 51228910, 51229805],\n",
       "      dtype=int32)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>a0</span></div><div class='xr-var-dims'>(variant)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;T&#x27; &#x27;A&#x27; &#x27;C&#x27; &#x27;C&#x27; ... &#x27;T&#x27; &#x27;A&#x27; &#x27;G&#x27; &#x27;T&#x27;</div><input id='attrs-3ccf328b-4b31-47c6-a53b-cc96c4ebae7b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3ccf328b-4b31-47c6-a53b-cc96c4ebae7b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d2aab7ed-de12-4af0-b8b1-c74283942be8' class='xr-var-data-in' type='checkbox'><label for='data-d2aab7ed-de12-4af0-b8b1-c74283942be8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;T&#x27;, &#x27;A&#x27;, &#x27;C&#x27;, ..., &#x27;A&#x27;, &#x27;G&#x27;, &#x27;T&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>a1</span></div><div class='xr-var-dims'>(variant)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;C&#x27; &#x27;C&#x27; &#x27;T&#x27; &#x27;T&#x27; ... &#x27;C&#x27; &#x27;G&#x27; &#x27;A&#x27; &#x27;C&#x27;</div><input id='attrs-cd325edb-f9b6-4648-8558-70469eb94f61' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-cd325edb-f9b6-4648-8558-70469eb94f61' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-cf0553df-350e-4ad5-ab2b-d923c46aa1a6' class='xr-var-data-in' type='checkbox'><label for='data-cf0553df-350e-4ad5-ab2b-d923c46aa1a6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;C&#x27;, &#x27;C&#x27;, &#x27;T&#x27;, ..., &#x27;G&#x27;, &#x27;A&#x27;, &#x27;C&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-119631fe-c78a-40e2-ad3d-a40bd7dfc864' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-119631fe-c78a-40e2-ad3d-a40bd7dfc864' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.DataArray 'genotype' (sample: 1610, variant: 131988)>\n",
       "dask.array<where, shape=(1610, 131988), dtype=float32, chunksize=(1024, 1024), chunktype=numpy.ndarray>\n",
       "Coordinates:\n",
       "  * sample   (sample) object 'HPSI0216i-tony_5' ... 'HPSI0416pf-cihd'\n",
       "  * variant  (variant) object '22_22_16051249_T_C' ... '22_22_51229805_T_C'\n",
       "    fid      (sample) object '0' '0' '0' '0' '0' '0' ... '0' '0' '0' '0' '0' '0'\n",
       "    iid      (sample) object 'HPSI0216i-tony_5' ... 'HPSI0416pf-cihd'\n",
       "    father   (sample) object '0' '0' '0' '0' '0' '0' ... '0' '0' '0' '0' '0' '0'\n",
       "    mother   (sample) object '0' '0' '0' '0' '0' '0' ... '0' '0' '0' '0' '0' '0'\n",
       "    gender   (sample) object '0' '0' '0' '0' '0' '0' ... '0' '0' '0' '0' '0' '0'\n",
       "    trait    (sample) object '-9' '-9' '-9' '-9' '-9' ... '-9' '-9' '-9' '-9'\n",
       "    chrom    (variant) object '22' '22' '22' '22' '22' ... '22' '22' '22' '22'\n",
       "    snp      (variant) object '22_16051249_T_C' ... '22_51229805_T_C'\n",
       "    cm       (variant) float64 0.0 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0 0.0\n",
       "    pos      (variant) int32 16051249 16051453 16053862 ... 51228910 51229805\n",
       "    a0       (variant) object 'T' 'A' 'C' 'C' 'G' 'C' ... 'G' 'T' 'A' 'G' 'T'\n",
       "    a1       (variant) object 'C' 'C' 'T' 'T' 'T' 'T' ... 'A' 'C' 'G' 'A' 'C'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "G_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nfs/software/stegle/users/acuomo/conda-envs/struct-lmm2/lib/python3.8/site-packages/xarray/core/indexing.py:1369: PerformanceWarning: Slicing with an out-of-order index is generating 2383 times more chunks\n",
      "  return self.array[key]\n",
      "/nfs/software/stegle/users/acuomo/conda-envs/struct-lmm2/lib/python3.8/site-packages/xarray/core/indexing.py:1369: PerformanceWarning: Slicing is producing a large chunk. To accept the large\n",
      "chunk and silence this warning, set the option\n",
      "    >>> with dask.config.set(**{'array.slicing.split_large_chunks': False}):\n",
      "    ...     array[indexer]\n",
      "\n",
      "To avoid creating the large chunks, set the option\n",
      "    >>> with dask.config.set(**{'array.slicing.split_large_chunks': True}):\n",
      "    ...     array[indexer]\n",
      "  return self.array[key]\n"
     ]
    }
   ],
   "source": [
    "# select down to relevant individuals\n",
    "G_exp = G_sel.sel(sample=sample_mapping[\"genotype_individual_id\"].values)\n",
    "assert all(L_expanded.sample.values == G_exp.sample.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get eigendecomposition of EEt\n",
    "[U, S, _] = economic_svd(E)\n",
    "us = U * S\n",
    "# get decomposition of K*EEt\n",
    "Ls = [ddot(us[:,i], L_expanded) for i in range(us.shape[1])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33964, 125)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ls[1].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>21843_1#10</th>\n",
       "      <th>21843_1#100</th>\n",
       "      <th>21843_1#101</th>\n",
       "      <th>21843_1#102</th>\n",
       "      <th>21843_1#103</th>\n",
       "      <th>21843_1#105</th>\n",
       "      <th>21843_1#106</th>\n",
       "      <th>21843_1#107</th>\n",
       "      <th>21843_1#108</th>\n",
       "      <th>21843_1#109</th>\n",
       "      <th>...</th>\n",
       "      <th>24539_8#88</th>\n",
       "      <th>24539_8#89</th>\n",
       "      <th>24539_8#90</th>\n",
       "      <th>24539_8#91</th>\n",
       "      <th>24539_8#92</th>\n",
       "      <th>24539_8#93</th>\n",
       "      <th>24539_8#94</th>\n",
       "      <th>24539_8#95</th>\n",
       "      <th>24539_8#97</th>\n",
       "      <th>24539_8#98</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003_TSPAN6</th>\n",
       "      <td>5.520777</td>\n",
       "      <td>6.456208</td>\n",
       "      <td>5.878671</td>\n",
       "      <td>4.860824</td>\n",
       "      <td>5.903640</td>\n",
       "      <td>4.513537</td>\n",
       "      <td>6.401983</td>\n",
       "      <td>5.909216</td>\n",
       "      <td>5.366645</td>\n",
       "      <td>3.228852</td>\n",
       "      <td>...</td>\n",
       "      <td>5.841814</td>\n",
       "      <td>6.104105</td>\n",
       "      <td>6.275649</td>\n",
       "      <td>7.029407</td>\n",
       "      <td>5.806978</td>\n",
       "      <td>6.199875</td>\n",
       "      <td>7.014180</td>\n",
       "      <td>6.228476</td>\n",
       "      <td>6.217161</td>\n",
       "      <td>6.034232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419_DPM1</th>\n",
       "      <td>5.392461</td>\n",
       "      <td>6.065923</td>\n",
       "      <td>6.838769</td>\n",
       "      <td>6.614268</td>\n",
       "      <td>6.512403</td>\n",
       "      <td>5.527439</td>\n",
       "      <td>6.525591</td>\n",
       "      <td>6.381135</td>\n",
       "      <td>6.157296</td>\n",
       "      <td>6.248478</td>\n",
       "      <td>...</td>\n",
       "      <td>6.543807</td>\n",
       "      <td>6.369119</td>\n",
       "      <td>7.185421</td>\n",
       "      <td>6.337047</td>\n",
       "      <td>6.162437</td>\n",
       "      <td>5.885993</td>\n",
       "      <td>7.431358</td>\n",
       "      <td>7.013124</td>\n",
       "      <td>4.851771</td>\n",
       "      <td>4.937248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457_SCYL3</th>\n",
       "      <td>0.000174</td>\n",
       "      <td>0.352597</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.825955</td>\n",
       "      <td>2.201697</td>\n",
       "      <td>0.262446</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.506837</td>\n",
       "      <td>0.283516</td>\n",
       "      <td>3.241977</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017386</td>\n",
       "      <td>0.949668</td>\n",
       "      <td>0.035526</td>\n",
       "      <td>0.032044</td>\n",
       "      <td>1.773369</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.108025</td>\n",
       "      <td>1.756339</td>\n",
       "      <td>2.492943</td>\n",
       "      <td>1.363441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460_C1orf112</th>\n",
       "      <td>1.471928</td>\n",
       "      <td>4.536968</td>\n",
       "      <td>4.318528</td>\n",
       "      <td>5.373009</td>\n",
       "      <td>4.636175</td>\n",
       "      <td>4.225468</td>\n",
       "      <td>0.409785</td>\n",
       "      <td>3.668277</td>\n",
       "      <td>3.057933</td>\n",
       "      <td>3.154891</td>\n",
       "      <td>...</td>\n",
       "      <td>4.720967</td>\n",
       "      <td>3.791536</td>\n",
       "      <td>2.696476</td>\n",
       "      <td>4.227515</td>\n",
       "      <td>4.243689</td>\n",
       "      <td>3.227508</td>\n",
       "      <td>2.621121</td>\n",
       "      <td>3.950978</td>\n",
       "      <td>3.926914</td>\n",
       "      <td>4.211904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001036_FUCA2</th>\n",
       "      <td>2.908802</td>\n",
       "      <td>3.867327</td>\n",
       "      <td>3.321747</td>\n",
       "      <td>3.736476</td>\n",
       "      <td>4.917576</td>\n",
       "      <td>2.456866</td>\n",
       "      <td>0.577839</td>\n",
       "      <td>4.777404</td>\n",
       "      <td>2.873857</td>\n",
       "      <td>2.536708</td>\n",
       "      <td>...</td>\n",
       "      <td>3.070608</td>\n",
       "      <td>4.491643</td>\n",
       "      <td>4.206249</td>\n",
       "      <td>3.695005</td>\n",
       "      <td>2.652845</td>\n",
       "      <td>4.221847</td>\n",
       "      <td>3.188030</td>\n",
       "      <td>4.741496</td>\n",
       "      <td>3.872743</td>\n",
       "      <td>4.374577</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34256 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          21843_1#10  21843_1#100  21843_1#101  21843_1#102  \\\n",
       "ENSG00000000003_TSPAN6      5.520777     6.456208     5.878671     4.860824   \n",
       "ENSG00000000419_DPM1        5.392461     6.065923     6.838769     6.614268   \n",
       "ENSG00000000457_SCYL3       0.000174     0.352597     0.000000     0.825955   \n",
       "ENSG00000000460_C1orf112    1.471928     4.536968     4.318528     5.373009   \n",
       "ENSG00000001036_FUCA2       2.908802     3.867327     3.321747     3.736476   \n",
       "\n",
       "                          21843_1#103  21843_1#105  21843_1#106  21843_1#107  \\\n",
       "ENSG00000000003_TSPAN6       5.903640     4.513537     6.401983     5.909216   \n",
       "ENSG00000000419_DPM1         6.512403     5.527439     6.525591     6.381135   \n",
       "ENSG00000000457_SCYL3        2.201697     0.262446     0.000000     1.506837   \n",
       "ENSG00000000460_C1orf112     4.636175     4.225468     0.409785     3.668277   \n",
       "ENSG00000001036_FUCA2        4.917576     2.456866     0.577839     4.777404   \n",
       "\n",
       "                          21843_1#108  21843_1#109  ...  24539_8#88  \\\n",
       "ENSG00000000003_TSPAN6       5.366645     3.228852  ...    5.841814   \n",
       "ENSG00000000419_DPM1         6.157296     6.248478  ...    6.543807   \n",
       "ENSG00000000457_SCYL3        0.283516     3.241977  ...    0.017386   \n",
       "ENSG00000000460_C1orf112     3.057933     3.154891  ...    4.720967   \n",
       "ENSG00000001036_FUCA2        2.873857     2.536708  ...    3.070608   \n",
       "\n",
       "                          24539_8#89  24539_8#90  24539_8#91  24539_8#92  \\\n",
       "ENSG00000000003_TSPAN6      6.104105    6.275649    7.029407    5.806978   \n",
       "ENSG00000000419_DPM1        6.369119    7.185421    6.337047    6.162437   \n",
       "ENSG00000000457_SCYL3       0.949668    0.035526    0.032044    1.773369   \n",
       "ENSG00000000460_C1orf112    3.791536    2.696476    4.227515    4.243689   \n",
       "ENSG00000001036_FUCA2       4.491643    4.206249    3.695005    2.652845   \n",
       "\n",
       "                          24539_8#93  24539_8#94  24539_8#95  24539_8#97  \\\n",
       "ENSG00000000003_TSPAN6      6.199875    7.014180    6.228476    6.217161   \n",
       "ENSG00000000419_DPM1        5.885993    7.431358    7.013124    4.851771   \n",
       "ENSG00000000457_SCYL3       0.000000    0.108025    1.756339    2.492943   \n",
       "ENSG00000000460_C1orf112    3.227508    2.621121    3.950978    3.926914   \n",
       "ENSG00000001036_FUCA2       4.221847    3.188030    4.741496    3.872743   \n",
       "\n",
       "                          24539_8#98  \n",
       "ENSG00000000003_TSPAN6      6.034232  \n",
       "ENSG00000000419_DPM1        4.937248  \n",
       "ENSG00000000457_SCYL3       1.363441  \n",
       "ENSG00000000460_C1orf112    4.211904  \n",
       "ENSG00000001036_FUCA2       4.374577  \n",
       "\n",
       "[5 rows x 34256 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Phenotype (genes X cells)\n",
    "phenotype_file = input_files_dir+\"phenotype.csv.pkl\"\n",
    "phenotype = pd.read_pickle(phenotype_file)\n",
    "phenotype.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phenotype shape BEFORE selection: (11231, 34256)\n",
      "Phenotype shape AFTER selection: (11231, 33964)\n"
     ]
    }
   ],
   "source": [
    "print(\"Phenotype shape BEFORE selection: {}\".format(phenotype.shape))\n",
    "phenotype = xr.DataArray(phenotype.values, dims=[\"trait\", \"cell\"], coords={\"trait\": phenotype.index.values, \"cell\": phenotype.columns.values})\n",
    "phenotype = phenotype.sel(cell=sample_mapping[\"phenotype_sample_id\"].values)\n",
    "print(\"Phenotype shape AFTER selection: {}\".format(phenotype.shape))\n",
    "assert all(phenotype.cell.values == sample_mapping[\"phenotype_sample_id\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>feature</th>\n",
       "      <th>stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5_149826526_C_T</td>\n",
       "      <td>ENSG00000164587_RPS14</td>\n",
       "      <td>ips</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11_57283988_C_T</td>\n",
       "      <td>ENSG00000134809_TIMM10</td>\n",
       "      <td>ips</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12_56401085_G_A</td>\n",
       "      <td>ENSG00000197728_RPS26</td>\n",
       "      <td>ips</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17_79634162_T_G</td>\n",
       "      <td>ENSG00000214087_ARL16</td>\n",
       "      <td>ips</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6_31486901_T_C</td>\n",
       "      <td>ENSG00000198563_DDX39B</td>\n",
       "      <td>ips</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            snp_id                 feature stage\n",
       "0  5_149826526_C_T   ENSG00000164587_RPS14   ips\n",
       "1  11_57283988_C_T  ENSG00000134809_TIMM10   ips\n",
       "2  12_56401085_G_A   ENSG00000197728_RPS26   ips\n",
       "3  17_79634162_T_G   ENSG00000214087_ARL16   ips\n",
       "4   6_31486901_T_C  ENSG00000198563_DDX39B   ips"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter on specific gene-SNP pairs\n",
    "# eQTL from endodiff (ips+mesendo+defendo)\n",
    "endo_eqtl_file = input_files_dir+\"endodiff_eqtl_allconditions_FDR10pct.csv\"\n",
    "endo_eqtl = pd.read_csv(endo_eqtl_file, index_col = False)\n",
    "endo_eqtl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "endo_eqtl[\"chrom\"] = [int(i[:i.find(\"_\")]) for i in endo_eqtl[\"snp_id\"]]\n",
    "genes = endo_eqtl[endo_eqtl['chrom']==int(chrom)]['feature'].unique()\n",
    "# genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "88"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up model\n",
    "n_samples = phenotype.shape[1]\n",
    "M = ones((n_samples, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSG00000100058_CRYBB2P1'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pick one gene as example\n",
    "i=0\n",
    "trait_name = genes[i]\n",
    "trait_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running for gene ENSG00000100058_CRYBB2P1\n"
     ]
    }
   ],
   "source": [
    "# select SNPs for a given gene\n",
    "leads = endo_eqtl[endo_eqtl['feature']==trait_name]['snp_id'].unique()\n",
    "G_tmp = G_exp[:,G_exp['snp'].isin(leads)]\n",
    "print(\"Running for gene {}\".format(trait_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# quantile normalise y, E\n",
    "y = phenotype.sel(trait=trait_name)\n",
    "y = quantile_gaussianize(y)\n",
    "E = quantile_gaussianize(E)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# null model\n",
    "slmm2 = StructLMM2(y.values, M, E, Ls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2/2 [09:26<00:00, 283.28s/it]\n"
     ]
    }
   ],
   "source": [
    "# run interaction test for the SNPs\n",
    "pvals = slmm2.scan_interaction(G_tmp)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pv</th>\n",
       "      <th>variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>9.681856e-14</td>\n",
       "      <td>22_25845855_C_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>9.906107e-03</td>\n",
       "      <td>22_25924999_G_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom            pv          variant\n",
       "0    22  9.681856e-14  22_25845855_C_G\n",
       "1    22  9.906107e-03  22_25924999_G_A"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pv = pd.DataFrame({\"chrom\":G_tmp.chrom.values,\n",
    "                   \"pv\":pvals,\n",
    "                   \"variant\":G_tmp.snp.values})\n",
    "pv.head()\n",
    "# pv.to_csv(outfilename, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (struct-lmm2)",
   "language": "python",
   "name": "struct-lmm2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
